src/standard/nep171.rs

```diff
@@ -0,0 +1,542 @@
+pub trait Nep171ControllerInternal {
+    /// Root storage slot.
+    fn root() -> Slot<()> {
+        Slot::root(DefaultStorageKey::Nep171)
+    }
+
+    /// Storage slot for the owner of a token.
+    fn slot_token_owner(token_id: &TokenId) -> Slot<AccountId> {
+        Self::root().field(StorageKey::TokenOwner(token_id))
+    }
+}
+
+/// Non-public controller interface for NEP-171 implementations.
+pub trait Nep171Controller {
+    /// Transfer a token from `sender_id` to `receiver_id`.
+    fn transfer(
+        &mut self,
+        token_id: &[TokenId],
+        current_owner_id: AccountId,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        memo: Option<String>,
+    ) -> Result<(), Nep171TransferError>;
+
+    /// Mints a new token `token_id` to `owner_id`.
+    fn mint(
+        &mut self,
+        token_ids: &[TokenId],
+        new_owner_id: &AccountId,
+        memo: Option<String>,
+    ) -> Result<(), Nep171MintError>;
+
+    /// Burns tokens `token_ids` owned by `current_owner_id`.
+    fn burn(
+        &mut self,
+        token_ids: &[TokenId],
+        current_owner_id: &AccountId,
+        memo: Option<String>,
+    ) -> Result<(), Nep171BurnError>;
+
+    /// Burns tokens `token_ids` without checking the owners.
+    fn burn_unchecked(&mut self, token_ids: &[TokenId]) -> bool;
+
+    /// Returns the owner of a token, if it exists.
+    fn token_owner(&self, token_id: &TokenId) -> Option<AccountId>;
+}
+
+/// Transfer metadata generic over both types of transfer (`nft_transfer` and
+/// `nft_transfer_call`).
+#[derive(
+    Serialize, Deserialize, BorshSerialize, BorshDeserialize, PartialEq, Eq, Clone, Debug, Hash,
+)]
+pub struct Nep171Transfer {
+    /// Current owner account ID.
+    pub owner_id: AccountId,
+    /// Sending account ID.
+    pub sender_id: AccountId,
+    /// Receiving account ID.
+    pub receiver_id: AccountId,
+    /// Optional approval ID.
+    pub approval_id: Option<u64>,
+    /// Token ID.
+    pub token_id: TokenId,
+    /// Optional memo string.
+    pub memo: Option<String>,
+    /// Message passed to contract located at `receiver_id` in the case of `nft_transfer_call`.
+    pub msg: Option<String>,
+}
+
+/// Contracts may implement this trait to inject code into NEP-171 functions.
+///
+/// `T` is an optional value for passing state between different lifecycle
+/// hooks. This may be useful for charging callers for storage usage, for
+/// example.
+pub trait Nep171Hook<T = ()> {
+    /// Executed before a token transfer is conducted.
+    ///
+    /// May return an optional state value which will be passed along to the
+    /// following `after_transfer`.
+    fn before_nft_transfer(&mut self, _transfer: &Nep171Transfer) -> T;
+
+    /// Executed after a token transfer is conducted.
+    ///
+    /// Receives the state value returned by `before_transfer`.
+    fn after_nft_transfer(&mut self, _transfer: &Nep171Transfer, _state: T);
+}
```

Comment:
I don't think so. We can restrict `T: Default` and have a default impl return `Default::default()`, but I don't think we can provide a _default_ (overridable) implementation for a specific instantiation of `T`.

Reply:
The macro does provide a `no_hooks` flag that removes the requirement of manually implementing hooks.

Reply:
It would not be possible to override the implementation with custom logic:

```rust
trait MyTrait<T = ()> {
    fn my_func() -> T;
}

impl<T, U> MyTrait<T> for U where T: Default {
    fn my_func() -> T {
        Default::default()
    }
}

struct MyStruct;

impl MyTrait<()> for MyStruct {
    fn my_func() -> () {
        todo!()
    }
}
```

```txt
error[E0119]: conflicting implementations of trait `MyTrait` for type `MyStruct`
   --> workspaces-tests/tests/non_fungible_token.rs:476:1
    |
468 | impl<T, U> MyTrait<T> for U where T: Default {
    | --------------------------- first implementation here
...
476 | impl MyTrait<()> for MyStruct {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `MyStruct`

```